xapian-letor/tests/api_letor.cc

```diff
@@ -0,0 +1,187 @@
+    termgenerator.set_stemmer(Xapian::Stem("en"));
+    termgenerator.index_text("Lions, Tigers, Bears and Giraffes", 1, "S");
+    termgenerator.index_text("This paragraph talks about lions and tigers and "
+			     "bears (oh, my!). It mentions giraffes, "
+			     "but that's not really very important. Lions "
+			     "and tigers are big cats, so they must be really "
+			     "cuddly. Bears are famous for being cuddly, at "
+			     "least when they're teddy bears.", 1, "XD");
+    termgenerator.index_text("Lions, Tigers, Bears and Giraffes");
+    termgenerator.increase_termpos();
+    termgenerator.index_text("This paragraph talks about lions and tigers and "
+			     "bears (oh, my!). It mentions giraffes, "
+			     "but that's not really very important. Lions "
+			     "and tigers are big cats, so they must be really "
+			     "cuddly. Bears are famous for being cuddly, at "
+			     "least when they're teddy bears.");
+    db.add_document(doc);
+    doc.clear_terms();
+    termgenerator.index_text("Lions, Tigers and Bears", 1, "S");
+    termgenerator.index_text("This is the paragraph of interest. Tigers are "
+			     "massive beasts - I wouldn't want to meet a "
+			     "hungry one anywhere. Lions are scary even when "
+			     "lyin' down. Bears are scary even when bare. "
+			     "Together I suspect they'd be less scary, as the "
+			     "tigers, lions, and bears would all keep each "
+			     "other busy. On the other hand, bears don't live "
+			     "in the same continent as far as I know.", 1,
+			     "XD");
+    termgenerator.index_text("Lions, Tigers and Bears");
+    termgenerator.increase_termpos();
+    termgenerator.index_text("This is the paragraph of interest. Tigers are "
+			     "massive beasts - I wouldn't want to meet a "
+			     "hungry one anywhere. Lions are scary even when "
+			     "lyin' down. Bears are scary even when bare. "
+			     "Together I suspect they'd be less scary, as the "
+			     "tigers, lions, and bears would all keep each "
+			     "other busy. On the other hand, bears don't live "
+			     "in the same continent as far as I know.");
+    db.add_document(doc);
+}
+
+DEFINE_TESTCASE(ranker1, generated) {
+    Xapian::Ranker *ranker = new Xapian::ListNETRanker();
+    string a = get_database_path("apitest_ranker1",
+				       gen_uniqterms_gt_doclen_db);
+    Xapian::Enquire enquire((Xapian::Database(a)));
+    enquire.set_query(Xapian::Query("lions"));
+    Xapian::MSet mymset = enquire.get_mset(0, 10);
+    string query ="20001 'lions'";
+    string qrel = "20001 Q0 1 1\n20001 Q0 2 2";
+    std::ofstream myfile;
+    myfile.open("query.txt");
+    myfile << query;
+    myfile.close();
+    myfile.open("qrel.txt");
+    myfile << qrel;
+    myfile.close();
+    Xapian::prepare_training_file(a, "query.txt", "qrel.txt", 10);
+    ranker->set_database_path(a);
+    ranker->set_query(Xapian::Query("lions"));
+    ranker->train_model("training-data.txt", "");
```

Comment:
You seem to be relying on `training-data.txt` being implicitly generated by the code above, yet name it explicitly here.

As I've said before (but was seemingly ignored) having a default name for this file is not helpful, and I think this illustrates some reasons why.  It's unclear to the reader where this file is coming from, and the default filename causes a collision if this testcase and the next are run in parallel - both testcases will try to write out the same file and then read it back, which will cause various erratic failures depending on the exact scheduling of the two tasks.

We should scrap the default and require the user to explicitly specify the filename to use.  And testcases should specify a filename such that there's no collision (e.g. by including the testcase name in the training data filename).